### PR TITLE
2756 jndi in observer

### PR DESCRIPTION
**Pull request by benjamin-confino**

### CODE REVIEW

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
Build https://wasrtc.hursley.ibm.com:9443/jazz/resource/itemOid/com.ibm.team.build.BuildResult/_3c5LIDcXEeiflZsT47ItAA was green!

File: dev/com.ibm.ws.cdi.jndi_fat/fat/src/com/ibm/ws/cdi12/fat/tests/JNDILookupTest.java [hunk 1]
```diff
@@ -31,11 +33,21 @@
 
     @BuildShrinkWrap
     public static Archive<?> buildShrinkWrap() {
+
+        JavaArchive jar = ShrinkWrap.create(JavaArchive.class, "jndiLookup.jar"); 
+        jar.addClass("com.ibm.ws.cdi12.test.jndi.observer.ObserverBean");
+
         WebArchive war = ShrinkWrap.create(WebArchive.class, "jndiLookup.war");
         war.add(new FileAsset(new File("test-applications/jndiLookup.war/resources/META-INF/permissions.xml")), "/META-INF/permissions.xml");
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.jndi_fat/fat/src/com/ibm/ws/cdi12/fat/tests/JNDILookupTest.java]: I think you need to move this permissions.xml into the new ear

File: dev/com.ibm.ws.cdi.jndi_fat/test-applications/jndiLookup.jar/resources/META-INF/permissions.xml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.jndi_fat/test-applications/jndiLookup.jar/resources/META-INF/permissions.xml]: I think this file is redundant ... you need it in the ear

File: dev/com.ibm.ws.cdi.jndi_fat/test-applications/jndiLookup.jar/src/com/ibm/ws/cdi12/test/jndi/observer/ObserverBean.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
+/*******************************************************************************
+ * Copyright (c) 2015 IBM Corporation and others.
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.jndi_fat/test-applications/jndiLookup.jar/src/com/ibm/ws/cdi12/test/jndi/observer/ObserverBean.java]: wrong date

File: dev/com.ibm.ws.cdi.jndi_fat/test-applications/jndiLookup.jar/src/com/ibm/ws/cdi12/test/jndi/observer/ObserverBean.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    public String getResult() {
+        return result;
+    }
+
+    public void init(@Observes @Initialized(ApplicationScoped.class)Object event){
+        try {
+            result = (String)new InitialContext().lookup("java:app/env/com.ibm.ws.cdi.jndi.test.result");
+        } catch (NamingException e) {
+            result = "test failed";
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.jndi_fat/test-applications/jndiLookup.jar/src/com/ibm/ws/cdi12/test/jndi/observer/ObserverBean.java]: wouldn't it be useful to know the exception message when it fails?

File: dev/com.ibm.ws.cdi.jndi_fat/test-applications/jndiLookup.war/src/com/ibm/ws/cdi12/test/jndi/LookupServlet.java [hunk 5]
```diff
@@ -18,17 +18,24 @@
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import com.ibm.ws.cdi12.test.jndi.observer.ObserverBean;
+
+import javax.enterprise.inject.spi.CDI;
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.jndi_fat/test-applications/jndiLookup.war/src/com/ibm/ws/cdi12/test/jndi/LookupServlet.java]: what is this import for?

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java [hunk 6]
```diff
@@ -373,20 +373,41 @@ public void applicationStarting(ApplicationInfo appInfo) throws StateChangeExcep
...
+            getCDIContainer().finalizeApplicationStarting(application);//This split is just to keep the CDIContainerImpl code conistant across liberty & websphere. 
         } catch (CDIException e) {
             throw new StateChangeException(e);
         } finally {
             if (oldCl != null) {
                 CDIUtils.getAndSetLoader(oldCl);
             }
+            if (setContext) {
+                ComponentMetaDataAccessorImpl accessor = ComponentMetaDataAccessorImpl.getComponentMetaDataAccessor();
+                accessor.endContext();
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java]: call the existing endContext() method

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java [hunk 7]
```diff
@@ -373,20 +373,41 @@ public void applicationStarting(ApplicationInfo appInfo) throws StateChangeExcep
...
+            //that didn't work so I use give classic utils one of the module archives.  
+
+            if (application.getModuleArchives().size() > 0 && 
+                    application.getApplicationMetaData() != null) {
+                CDIArchive archive = application.getModuleArchives().iterator().next();
+                beginContext(archive);
+                setContext = true;
+        }
             getCDIContainer().applicationStarting(application);
+            getCDIContainer().finalizeApplicationStarting(application);//This split is just to keep the CDIContainerImpl code conistant across liberty & websphere.
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java]: change to `startInitialization` and `endInitialization`

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java [hunk 8]
```diff
@@ -373,20 +373,41 @@ public void applicationStarting(ApplicationInfo appInfo) throws StateChangeExcep
...
+            	
+            //Origonally I tried to setup JNDI so only application metadata was on the thread but
+            //that didn't work so I use give classic utils one of the module archives.  
+
+            if (application.getModuleArchives().size() > 0 && 
+                    application.getApplicationMetaData() != null) {
+                CDIArchive archive = application.getModuleArchives().iterator().next();
+                beginContext(archive);
+                setContext = true;
+        }
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java]: indentation is wrong ... did you use a tab somehow?

File: dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java [hunk 9]
```diff
@@ -15,9 +15,11 @@
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.HashMap;
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java]: remove unused imports

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Pull request merged**